### PR TITLE
feat: add mypy as type checker, update code and add comment to reference

### PR DESCRIPTION
**Pull request by pythonbyte**
This PR has the intention to add mypy as the type-checker and also update the GH actions to reflect that. 

Along with that, some fixes were done on the typing, and comments were added to type:ignore along with the warning that we need to fix/improve in the future.

### CODE REVIEW

File: .github/workflows/type-checker.yml [hunk 1]
```diff
@@ -12,19 +11,16 @@ jobs:
...
         run: |
-          sudo apt-get update &&
-          pip install poetry &&
-          poetry lock &&
-          poetry install
+          pip install mypy
 
       - name: Run type checks
-        run: poetry run pyright
+        run: mypy src --install-types --non-interactive --ignore-missing-imports --exclude "cli/templates"
```
Comment (gvieira) [inline .github/workflows/type-checker.yml]: Any way we can setup such flags in a conf file somewhere, like pyproject.toml?

File: src/crewai/memory/memory.py [hunk 2]
```diff
@@ -12,12 +12,16 @@ def __init__(self, storage: Storage):
         self.storage = storage
 
     def save(
-        self, value: Any, metadata: Dict[str, Any] = None, agent: str = None
+        self,
+        value: Any,
+        metadata: Optional[Dict[str, Any]] = None,
+        agent: Optional[str] = None,
```
Comment (gvieira) [inline src/crewai/memory/memory.py:18]: What is your take on `Optional[str]` vs `str | None`? I've been using the latter these days.

File: .github/workflows/type-checker.yml (hunk 1 above)
Comment (pythonbyte) [reply .github/workflows/type-checker.yml]: Yes of course, this was just the first approach to see if it worked

File: src/crewai/memory/memory.py (hunk 2 above)
Comment (pythonbyte) [reply src/crewai/memory/memory.py:18]: To be honest not a strong opinion, I kinda like the direct "Optional" making it clear that is a Str/None. 
But I saw that recently was changed to <type> | None. 

I can change to the <type> | None so we keep a standard throughout the project

Comment (gvieira) [reply src/crewai/memory/memory.py:18]: Yeah. If `| None` is the updated way to do it, let's try to use it. But there's no reason to start switching everywhere. We can adjust as we move forward.

**Pull request merged**